serde_arrow/src/internal/schema/tracer.rs

```diff
@@ -965,18 +965,46 @@ impl UnionTracer {
             }
         }
 
-        let mut fields = Vec::new();
-        for (idx, variant) in self.variants.iter().enumerate() {
-            if let Some(variant) = variant {
-                fields.push((i8::try_from(idx)?, variant.tracer.to_field()?));
-            } else {
-                fields.push((i8::try_from(idx)?, unknown_variant_field()));
-            };
+        let data_type: DataType;
+
+        if self.options.enums_with_data_as_structs {
+            let mut fields = Vec::new();
+
+            // For this option, we want to merge the variant children up one level, combining the names
+            // For each variant with name variant_name
+            // For each variant_field with field_name
+            // Add field {variant_name}_{field_name} -> variant_field.to_field() that is nullable
+
+            for variant in &self.variants {
+                if let Some(variant) = variant {
+                    // TODO: does this break if there are no child fields?
+                    let schema = variant.tracer.to_schema()?;
+                    for mut field in schema.fields {
+                        field.name = format!("{}_{}", variant.name.to_lowercase(), field.name);
```

Comment:
I like your suggestion about using a separator that is not valid Rust code to guarantee certain cases cannot arise.

Reply:
> What happens with only nullable fields in variant? (I would throw an error)
> * Most likely your deserialization logic will look for any-null field and select the matching variant. However, if there is a variant for which all fields are `None` this logic will fail.

For this case, I believe it's best to continue to serialize it. I agree at first we simply don't add support for deserialization and it seems perfectly fine for the user to serialize out an enum where the variant has all nullable fields. If we want to indicate an error, we can check all fields null when deserializing (whenever we get there) and raise the error at that point.

Reply:
Hi @chmp , just to clarify on the points you made and make sure nothing is missed

> * What happens if there is a conflict between unions (see below)?

I have changed the implementation to use the `::` separator which I believe addresses the chance of a inadvertent naming collision.

> * What happens with unions without fields (I would throw an error, as it will be hard to detect the correct variant on deserialization)?

This currently fails in the `to_field()` method of the UnionTracer when it detects enums without data.
 
> * What happens with non-struct variants (unit variants, tuple variants, new-type variants?) (I would probably forbid unit variants, as they fall under the no-fields rule)

- unit variants: if `enums_without_data_as_strings()` is enabled, fields are serialized using that approach. if not enabled, it throws an error, as previously done.
- tuple variants: This currently re-raises the error that occurs when `to_schema` is called on the variant tracer in the `UnionTracer::to_field()` method. I believe this maintains the existing behavior?
- new-type variants: this should be exactly the same as tuple structs, right? I believe it's handled in the same way?

> * What happens with only nullable fields in variant? (I would throw an error)

I commented on this separately above

> It would be great if you could add tests for these corner cases.

I added some tests confirming what I hope you agree is the expected behavior but if I misunderstood anything, please feel free to point it out. Thanks!